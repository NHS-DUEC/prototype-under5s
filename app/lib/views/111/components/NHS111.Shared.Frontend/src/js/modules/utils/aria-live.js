export function ariaAssertiveMessage(message) {
  const ariaAssertiveRegion = document.getElementById("aria-assertive");
  const el = document.createElement("p");
  el.textContent = message;
  ariaAssertiveRegion.appendChild(el);

  setTimeout(() => {
    el.remove();
  }, 100);
}

export function ariaPoliteMessage(message) {
  const ariaPoliteRegion = document.getElementById("aria-polite");
  const el = document.createElement("p");
  el.textContent = message;
  ariaPoliteRegion.appendChild(el);

  setTimeout(() => {
    el.remove();
  }, 100);
}
