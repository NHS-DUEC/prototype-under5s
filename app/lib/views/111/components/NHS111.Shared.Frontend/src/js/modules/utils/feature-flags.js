const featureFlags = JSON.parse(
  document.querySelector("#feature-flags").innerHTML
);

export function featureIsEnabled(featureName) {
  return featureFlags[featureName];
}
