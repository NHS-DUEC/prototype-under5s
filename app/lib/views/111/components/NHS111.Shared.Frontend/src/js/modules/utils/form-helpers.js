export function hideNoJsMessage(form) {
  Array.from(form.querySelectorAll("[data-feedback-nojs-message]")).forEach(
    (item) => item.remove()
  );
}

export function clearErrors(form) {
  const errors = form.querySelectorAll("[data-feedback-error]");
  const formGroups = form.querySelectorAll(".nhsuk-form-group");
  const elementsWithOriginalAria = form.querySelectorAll(
    "[data-original-aria-describedby]"
  );

  Array.from(errors).forEach((item) => item && item.remove());
  Array.from(formGroups).forEach(
    (item) => item && item.classList.remove("nhsuk-form-group--error")
  );
  Array.from(elementsWithOriginalAria).forEach((el) =>
    el.setAttribute("aria-describedBy", el.dataset.originalAriaDescribedby)
  );
}

export function hideAllFormSections(form, formSections) {
  clearErrors(form);

  Array.from(formSections).forEach((element) => {
    element.hidden = true;
  });
}

export function trackOriginalAria(form) {
  // Keep track of any original aria-describedBy attributes so we can restore them when clearing errors
  Array.from(form.querySelectorAll("fieldset, input, textarea")).forEach(
    (el) => {
      if (el.getAttribute("aria-describedBy") !== null) {
        el.setAttribute(
          "data-original-aria-describedby",
          el.getAttribute("aria-describedBy")
        );
      }
    }
  );
}

export function formError(
  form,
  ErrorSummary,
  fieldId,
  errorMessage,
  errorLink = ""
) {
  const template = document.createElement("template");

  clearErrors(form);

  template.innerHTML = `
    <div class="nhsuk-error-summary" data-feedback-error data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#${fieldId}">${errorLink || errorMessage}</a>
          </li>
        </ul>
      </div>
    </div>`;

  const previousErrorSummary = form.querySelector(".nhsuk-error-summary");
  if (previousErrorSummary) {
    previousErrorSummary.remove();
  }

  const errorSummary = template.content;

  form.prepend(errorSummary);
  ErrorSummary();

  const field = form.querySelector(`#${fieldId}`);
  const formGroup = field.closest(".nhsuk-form-group");

  const errorMessageTemplate = document.createElement("template");
  errorMessageTemplate.innerHTML = `<span data-feedback-error data-test-id="feedback-error-message" class="nhsuk-error-message" id="${fieldId}-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> ${errorMessage}
    </span>`;

  if (field.getAttribute("type") === "radio") {
    formGroup.querySelector("legend").after(errorMessageTemplate.content);
    const fieldset = formGroup.querySelector("fieldset");
    fieldset.setAttribute(
      "aria-describedby",
      `${
        fieldset.dataset.originalAriaDescribedby
          ? fieldset.dataset.originalAriaDescribedby
          : ""
      } ${fieldId}-error`
    );
  } else {
    formGroup.querySelector("label").after(errorMessageTemplate.content);
    field.setAttribute(
      "aria-describedby",
      `${
        field.dataset.originalAriaDescribedby
          ? field.dataset.originalAriaDescribedby
          : ""
      } ${fieldId}-error`
    );
  }

  formGroup.classList.add("nhsuk-form-group--error");
}
