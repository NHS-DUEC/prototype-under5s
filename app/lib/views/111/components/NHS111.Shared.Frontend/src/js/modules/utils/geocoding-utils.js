const addressStringToLatLongCache = {};

export async function addressStringToLatLong(
  GoogleMapsGeocoder,
  addressString
) {
  return new Promise((resolve, reject) => {
    if (addressString in addressStringToLatLongCache) {
      resolve(addressStringToLatLongCache[addressString]);
    } else {
      const geocoder = new GoogleMapsGeocoder();
      geocoder.geocode(
        { address: addressString, componentRestrictions: { country: "GB" } },
        (data) => {
          addressStringToLatLongCache[addressString] =
            data[0].geometry.location;
          resolve(addressStringToLatLongCache[addressString]);
        }
      );
    }
  });
}
