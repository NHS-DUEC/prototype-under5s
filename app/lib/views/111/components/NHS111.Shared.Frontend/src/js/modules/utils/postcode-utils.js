function lastIndexOfDigit(str) {
  const lastIndices = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map((digit) =>
    str.lastIndexOf(digit)
  );
  const indices = lastIndices.sort((a, b) => {
    return a - b;
  });
  return indices[indices.length - 1];
}

export function getPartPostCode(postcode) {
  const postcodeWithoutWhitespace = postcode.replace(/\s/g, "");

  if (postcodeWithoutWhitespace.length - 3 <= 1)
    return postcodeWithoutWhitespace;

  const lastDigit = lastIndexOfDigit(postcodeWithoutWhitespace);

  if (lastDigit === -1) {
    return postcodeWithoutWhitespace;
  }

  if (lastDigit < postcodeWithoutWhitespace.length - 1)
    return postcodeWithoutWhitespace.substring(0, lastDigit);

  return postcodeWithoutWhitespace;
}
