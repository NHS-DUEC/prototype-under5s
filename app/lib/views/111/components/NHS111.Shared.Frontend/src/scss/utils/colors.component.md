---
name: colors
---

Utility classes for colours.

<div class="app-u-color-red">
  <pre>class="app-u-color-red"</pre>
</div>

<hr>

<div class="app-u-color-secondary-text">
  <pre>class="app-u-color-secondary-text"</pre>
</div>

<hr>

<div class="app-u-background-pink">
  <pre>class="app-u-background-pink"</pre>
</div>
