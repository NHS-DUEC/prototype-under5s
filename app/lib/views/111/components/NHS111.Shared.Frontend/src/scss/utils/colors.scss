.app-u-color-red {
  color: $color_nhsuk-red;
}

.app-u-color-secondary-text {
  color: #415462;
}

.app-u-background-pink {
  background-color: #fae9e8;
}

.app-u-background-white {
  background-color: $color_nhsuk-white;
}
