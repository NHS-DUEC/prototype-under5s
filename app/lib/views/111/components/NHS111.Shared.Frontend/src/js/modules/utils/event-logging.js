import { getPartPostCode } from "./postcode-utils";

export function getHiddenFieldValue(name, nameEndsWith = false) {
  const fields = nameEndsWith
    ? document.querySelectorAll(`[name$="${name}"]`)
    : document.getElementsByName(name);

  if (fields.length > 0) {
    return fields[0].value;
  }

  return "";
}

export function getGender() {
  const gender = getHiddenFieldValue("Gender", true);

  if (["male", "female"].includes(gender.toLowerCase())) return gender;

  return "";
}

export function getDxCode() {
  // find the DxCode by searching for any input with value starting Dx
  const dxCode = document.querySelector('input[value^="Dx"]');

  if (dxCode) {
    return dxCode.value;
  }

  return "";
}

export function createAuditEntry(
  eventKey,
  eventValue,
  eventData = "",
  answerOrder = "",
  answerTitle = ""
) {
  return {
    journeyId: getHiddenFieldValue("JourneyId", true),
    campaign: getHiddenFieldValue("Campaign"),
    dxCode: getDxCode(),
    postCodePart: getPartPostCode(getHiddenFieldValue("CurrentPostcode", true)),
    pathwayId: getHiddenFieldValue("PathwayId", true),
    pathwayTitle: getHiddenFieldValue("PathwayTitle", true),
    campaignSource: getHiddenFieldValue("Source"),
    QuestionNo: getHiddenFieldValue("QuestionNo"),
    QuestionTitle: getHiddenFieldValue("Title"),
    age: getHiddenFieldValue("Age", true),
    gender: getGender(),
    eventKey,
    eventValue,
    eventData,
    page: window.location.pathname,
    AnswerOrder: answerOrder,
    AnswerTitle: answerTitle,
  };
}

export function logEvent(
  eventKey,
  eventValue,
  eventData,
  answerOrder,
  answerTitle
) {
  const formData = new FormData();
  const auditEntry = createAuditEntry(
    eventKey,
    eventValue,
    eventData,
    answerOrder,
    answerTitle
  );

  Object.entries(auditEntry).forEach((item) => {
    const [key, value] = item;
    formData.append(key, value);
  });

  if (navigator.sendBeacon) {
    navigator.sendBeacon("/Auditing/Log", formData);
  }
}

export function logClick(eventValue, answerOrder, answerTitle) {
  logEvent("Clicked", eventValue, "", answerOrder, answerTitle);
}
