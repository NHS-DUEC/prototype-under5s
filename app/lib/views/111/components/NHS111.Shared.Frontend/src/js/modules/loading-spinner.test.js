jest.mock("./utils/aria-live");

import initLoadingSpinners from "./loading-spinner";
import { ariaPoliteMessage } from "./utils/aria-live";

const mockSubmit = jest.fn().mockName("Form submit");
const defaultSubmitImplementation = window.HTMLFormElement.prototype.submit;

describe("loading spinner", () => {
  beforeAll(() => {
    window.HTMLFormElement.prototype.submit = mockSubmit;
  });

  beforeEach(() => {
    mockSubmit.mockClear();
    ariaPoliteMessage.mockClear();
  });

  afterAll(() => {
    window.HTMLFormElement.prototype.submit = defaultSubmitImplementation;
  });

  describe("bound to a form", () => {
    test("shows loading overlay when submitted", (done) => {
      document.body.innerHTML = `<form id="test-form" data-loading-spinner="Overlay text">
        <button type="submit" id="test-button">Test</button>
      </form>
      `;

      initLoadingSpinners();

      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the overlay is shown
        expect(mockSubmit).toHaveBeenCalled();
        expect(mockSubmit.mock.calls.length).toBe(1);
        expect(ariaPoliteMessage).toHaveBeenCalledWith(
          "Please wait - Overlay text"
        );
        done();
      }, 100);
    });
  });

  describe("bound to a clickable element", () => {
    test("shows loading overlay when clicked", (done) => {
      document.body.innerHTML = `<button data-loading-spinner="Overlay text" id="test-button">Test</button>`;

      initLoadingSpinners();

      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the overlay is shown
        expect(ariaPoliteMessage).toHaveBeenCalledWith(
          "Please wait - Overlay text"
        );
        done();
      }, 0);
    });

    test("inline spinner is hidden again when pageshow event fires", (done) => {
      document.body.innerHTML = `<button data-loading-spinner="Overlay text" id="test-button">Test</button>`;

      initLoadingSpinners();

      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the overlay is shown
        expect(ariaPoliteMessage).toHaveBeenCalledWith(
          "Please wait - Overlay text"
        );

        window.dispatchEvent(new Event("pageshow"));
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the overlay is no longer shown
        done();
      }, 0);
    });

    test("inline spinner is hidden again when the custom hide-loading-spinner event fires", (done) => {
      document.body.innerHTML = `<button data-loading-spinner="Overlay text" id="test-button">Test</button>`;

      initLoadingSpinners();

      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the overlay is shown
        expect(ariaPoliteMessage).toHaveBeenCalledWith(
          "Please wait - Overlay text"
        );

        const event = new CustomEvent("hide-loading-spinner", {
          bubbles: true,
        });

        document.querySelector("#test-button").dispatchEvent(event);

        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the overlay is no longer shown
        done();
      }, 0);
    });

    test("loads custom data from hidden fields when shown", (done) => {
      document.body.innerHTML = `
      <input type="hidden" id="find-by-id" value="Hidden fields by ID" />
      <input type="hidden" name="find-by-name" value="Hidden fields by name" />
      <input type="hidden" data-attribute="find-by-data-attribute" value="Hidden fields by data attribute" />

      <button data-loading-spinner="by ID: {{#find-by-id}} | by name: {{[name=find-by-name]}} | by data attribute: {{[data-attribute]}}" id="test-button">Test</button>`;

      initLoadingSpinners();

      document.querySelector("#test-button").click();

      setTimeout(() => {
        expect(document.body.innerHTML).toMatchSnapshot(); // Assert that the overlay is shown
        expect(ariaPoliteMessage).toHaveBeenCalledWith(
          "Please wait - by ID: Hidden fields by ID | by name: Hidden fields by name | by data attribute: Hidden fields by data attribute"
        );
        done();
      }, 0);
    });
  });
});
