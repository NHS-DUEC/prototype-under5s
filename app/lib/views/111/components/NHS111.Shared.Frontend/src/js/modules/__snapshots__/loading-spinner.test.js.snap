// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner bound to a clickable element inline spinner is hidden again when pageshow event fires 1`] = `"<button data-loading-spinner="Overlay text" id="test-button">Test</button><div class="app-loading-overlay"><div class="app-loading-overlay__inner"><p class="app-loading-overlay__text">Overlay text</p><div class="app-spinner"><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div></div></div></div>"`;

exports[`loading spinner bound to a clickable element inline spinner is hidden again when pageshow event fires 2`] = `"<button data-loading-spinner="Overlay text" id="test-button">Test</button>"`;

exports[`loading spinner bound to a clickable element inline spinner is hidden again when the custom hide-loading-spinner event fires 1`] = `"<button data-loading-spinner="Overlay text" id="test-button">Test</button><div class="app-loading-overlay"><div class="app-loading-overlay__inner"><p class="app-loading-overlay__text">Overlay text</p><div class="app-spinner"><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div></div></div></div>"`;

exports[`loading spinner bound to a clickable element inline spinner is hidden again when the custom hide-loading-spinner event fires 2`] = `"<button data-loading-spinner="Overlay text" id="test-button">Test</button>"`;

exports[`loading spinner bound to a clickable element loads custom data from hidden fields when shown 1`] = `
"
      <input type="hidden" id="find-by-id" value="Hidden fields by ID">
      <input type="hidden" name="find-by-name" value="Hidden fields by name">
      <input type="hidden" data-attribute="find-by-data-attribute" value="Hidden fields by data attribute">

      <button data-loading-spinner="by ID: {{#find-by-id}} | by name: {{[name=find-by-name]}} | by data attribute: {{[data-attribute]}}" id="test-button">Test</button><div class="app-loading-overlay"><div class="app-loading-overlay__inner"><p class="app-loading-overlay__text">by ID: Hidden fields by ID | by name: Hidden fields by name | by data attribute: Hidden fields by data attribute</p><div class="app-spinner"><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div></div></div></div>"
`;

exports[`loading spinner bound to a clickable element shows loading overlay when clicked 1`] = `"<button data-loading-spinner="Overlay text" id="test-button">Test</button><div class="app-loading-overlay"><div class="app-loading-overlay__inner"><p class="app-loading-overlay__text">Overlay text</p><div class="app-spinner"><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div></div></div></div>"`;

exports[`loading spinner bound to a form shows loading overlay when submitted 1`] = `
"<form id="test-form" data-loading-spinner="Overlay text">
        <button type="submit" id="test-button">Test</button>
      </form><div class="app-loading-overlay"><div class="app-loading-overlay__inner"><p class="app-loading-overlay__text">Overlay text</p><div class="app-spinner"><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div></div></div></div>
      "
`;
